Crashlytics/Crashlytics/Rollouts/CrashlyticsRemoteConfigManager.swift

```diff
@@ -15,38 +15,56 @@
 import FirebaseRemoteConfigInterop
 import Foundation
 
-protocol CrashlyticsPersistentLog: NSObject {
-  func updateRolloutsStateToPersistence(rolloutAssignments: [RolloutAssignment])
+@objc(FIRCLSPersistenceLog)
+public protocol CrashlyticsPersistenceLog {
+  func updateRolloutsStateToPersistence(rollouts: Data, reportID: String)
 }
 
 @objc(FIRCLSRemoteConfigManager)
 public class CrashlyticsRemoteConfigManager: NSObject {
   public static let maxRolloutAssignments = 128
   public static let maxParameterValueLength = 256
 
+  private let lock = NSLock()
+  private var _rolloutAssignment: [RolloutAssignment] = []
+
   var remoteConfig: RemoteConfigInterop
-  @objc public private(set) var rolloutAssignment: [RolloutAssignment] = []
-  weak var persistenceDelegate: CrashlyticsPersistentLog?
+  var persistenceDelegate: CrashlyticsPersistenceLog
 
-  @objc public init(remoteConfig: RemoteConfigInterop) {
-    self.remoteConfig = remoteConfig
+  @objc public var rolloutAssignment: [RolloutAssignment] {
+    lock.lock()
+    defer { lock.unlock() }
+    let copy = _rolloutAssignment
```

Comment:
Keep in mind this will copy the array, but not the values of each RolloutAssignment (only the references): https://stackoverflow.com/questions/27812433/how-do-i-make-a-exact-duplicate-copy-of-an-array

If we're worried about those Rollout Assignments getting deleted (we might not be!), then we'll want to iterate and copy each of those too.

Reply:
I see, thank you for sharing the info! I think reference here should be fine as long as we have the strong reference of the assignment objects somewhere? The reason I have this logic here is preventing `_rolloutAssignment` gets overwritten while reading. If such thing happen the old assignment objects will deinit because no reference anymore(the ARC concept). But since here the copy still hold a reference so the old assignments will be used for reading.

Reply:
Added unit test for this!